src/Discord.Net.Core/Utils/Cacheable.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Discord
+{
+    /// <summary>
+    /// Contains an entity that may be cached.
+    /// </summary>
+    /// <typeparam name="TEntity">The type of entity that is cached</typeparam>
+    /// <typeparam name="TId">The type of this entity's ID</typeparam>
+    public struct Cacheable<TEntity, TId>
+        where TEntity : IEntity<TId>
+        where TId : IEquatable<TId>
+    {
+        /// <summary>
+        /// Is this entity cached?
+        /// </summary>
+        public bool HasValue { get; }
```

Comment:
No, because null is a valid value, and there's no class constraint. We can't be sure we wont use this helper in other places, so best to be safe.